lib/r10k/git/rugged/credentials.rb

```diff
@@ -43,7 +44,10 @@ def get_ssh_key_credentials(url, username_from_url)
 
     global_private_key = R10K::Git.settings[:private_key]
 
-    if per_repo_private_key
+    if @credentials_from_cli[:sshkey_file]
+      private_key = @credentials_from_cli[:sshkey_file]
+      logger.debug2 _("Using per-deploy private key %{key} for URL %{url}") % {key: private_key, url: url.inspect}
+    elsif per_repo_private_key
```

Comment:
So... we already allow folks to specify the the ssh key path in their configuration? Can we treat `--sshkey_path` as overriding whatever option was given in that file. Like how in puppet you can set any setting in the puppet.conf but then override it on the command line with `--<setting_name>`? Then overwrite the value with the cli supplied value in whatever `R10K::Git.get_repo_settings(url)` is looking up?

Reply:
Just to be clear, this is an alternative plumbing approach (using the settings subsystem instead of treating it as an option on the repo), right, not a desire for a functionality change?

Just asking because the per-repo stuff does not behave like this does. CLI settings apply to all repos, so it's more like an override for the global private key that takes precedence over everything in the config. And I'm not sure we have a choice about that, because sources are not exposed as an abstraction at the CLI level (but they are in the config).

Reply:
So I would say, yes, it's probably doable, and has the advantage of touching fewer places and not changing the API as much. We can probably add something on that same settings object for the token too, though we'd want to make sure there isn't anyway to print all the settings or something like that.

Reply:
I can plumb it through the settings architecture, which I think is better (fewer API changes), but I think I'll have to make a new setting for the private key to get the precedence to work right. Still seems like an improvement overall.

Reply:
See https://github.com/puppetlabs/r10k/pull/1116 for work on this approach